Move CSP code into one entrypoint

TimvdLippe opened the pull request:
This refactoring moves various CSP-related methods away from GlobalScope and Document into a dedicated entrypoint. It also reduces the amount of imports of the CSP crate, so that types are consolidated into this one entrypoint. That way, we control how CSP code interacts with the script crate.

For reviewing purposes, I split up the refactoring into separate distinct commits that all move 1 method(group) into the new file.

Testing: no change in behavior, only a build improvement + code cleanup

TimvdLippe committed:
Move report_csp_violations out of globalscope

It doesn't use any internal members, so it can be in a separate
contained file.

Signed-off-by: Tim van der Lippe <tvanderlippe@gmail.com>

TimvdLippe committed:
Move js_is_evaluation_allowed out of GlobalScope

Signed-off-by: Tim van der Lippe <tvanderlippe@gmail.com>

TimvdLippe committed:
Move should_navigation_request_be_blocked out of GlobalScope

Signed-off-by: Tim van der Lippe <tvanderlippe@gmail.com>

TimvdLippe committed:
Move parse_csp_list_from_metadata out of GlobalScope

Signed-off-by: Tim van der Lippe <tvanderlippe@gmail.com>

TimvdLippe committed:
Move should_elements_inline_type_behavior_be_blocked out of Document

Also change it to accept a GlobalScope, to keep it consistent with
the other methods. They refer to the same CSP List anyways.

Signed-off-by: Tim van der Lippe <tvanderlippe@gmail.com>

TimvdLippe committed:
Reduce number of direct imports of CSP crate

Rather than mix-matching the imports of CSP code, now almost
all usages of the CSP crate are changed to use the csp script
file. That way, we consolidate the usages of CSP and control
what is exposed to other files.

Signed-off-by: Tim van der Lippe <tvanderlippe@gmail.com>

TimvdLippe committed:
Move trusted type CSP checks out of trustedtypepolicyfactory

Signed-off-by: Tim van der Lippe <tvanderlippe@gmail.com>

TimvdLippe commented:
https://github.com/TimvdLippe/servo/actions/runs/15794076394

TimvdLippe committed:
Cleanup trusted type code

Return true instead of false and use less indentation for readability.

Signed-off-by: Tim van der Lippe <tvanderlippe@gmail.com>

mrobinson reviewed (changes requested):
Nice!

Can we take advantage of Rust's flexibility here and simply move these methods to `csp.rs` but keep them inside `impl GlobalScope` and `impl Document` blocks. In Rust, these can be spread across different code units. This will make calling them a bit nicer.

TimvdLippe commented on components/script/dom/document.rs:
```diff
@@ -4306,30 +4305,6 @@ impl Document {
     }
 
-    /// <https://www.w3.org/TR/CSP/#should-block-inline>
-    pub(crate) fn should_elements_inline_type_behavior_be_blocked(
```
@mrobinson Maybe to elaborate more: this method lived in Document, but it never should have. That's because while the specification talks about using documents to documents global csp list, that's the only usage of document in here. In fact, after moving I realized that it is inconsistent with the other CSP methods and it should take the GlobalScope instead.

Therefore, I think by separating the two, there is a clearer distinction in which code does what. The CSP code does use the GlobalScope csp_list, but that's it. There are no other real connections between the two, so I think decoupling as much as a possible is preferable.

mrobinson replied:
If these functions had to be methods on a struct, what struct would they be methods on? The main thing I'd like to get away from in script is having a bunch of free functions scattered throughout the code. Is it possible to turn these into methods?

TimvdLippe replied:
I understand that concern and I agree I would prefer these to be on a struct. That struct would be `CspList`, but that lives in the `content_security_policy` crate. As I understood it, it isn't possible to add more methods to a struct in a different crate, is it? Otherwise we could create a CSP struct, but then never interact with it other than directly statically calling the methods. I thought that wouldn't be better than having the methods like this in a single file. WDYT?

mrobinson replied:
One way we do this is to create a `trait` that is implemented by a single struct.  I think this would be a decent way to organize the code if it sounds good to you.

TimvdLippe replied:
Sure, I am not sure how that's different other than calling it via a function directly. As in, isn't the trait a copy-paste and doesn't get referenced anywhere? Could you point me to an example in the codebase so I understand better how to implement it? I feel like my lack of Rust knowledge is a factor here.

jdm replied:
Any cide that calls the trait methods on CSP list would need to `use` the trait. Not sure what you mean by copying and pasting.https://github.com/servo/servo/blob/f370606fa81851f2a401534adb5b91d7c456e4cc/components/fonts/platform/macos/font.rs#L356 is an example.

TimvdLippe replied:
Ah I see. I don't have experience implementing that, but eager to learn how it works. One caveat: `global.get_csp_list()` returns an `Option<CspList>`. Hence the unwrapping of the option and handling the default value is up to the caller if we use a trait. The reason I wrote these methods in the first place is to handle this complexity. It ultimately leads to worse callsides, since they now need to deal with the default value and I myself have made the mistake multiple times (including while working on this PR) where I mixed a `true` with a `false`.

One way out I see is to make `get_csp_list()` return a `CspList`, in line with #36828. Then, we can implement the trait here and handle it. The method names would get awkward, since currently we reuse the same method names (since they are wrappers). If we implement them on the trait, then we need to use a different name. E.g. it would become `is_wasm_evaluation_allowed_and_report`.

WDYT about that?

TimvdLippe replied:
Thanks, I did not know that was possible! I added two more commits to move the relevant methods to traits. PTAL

TimvdLippe committed:
Move CSP reporting methods to separate trait

Implemented on an Option<CspList> to handle the None case
in a single place, similar to the functions had before.

Signed-off-by: Tim van der Lippe <tvanderlippe@gmail.com>

TimvdLippe committed:
Move report_csp_violations to a trait

Signed-off-by: Tim van der Lippe <tvanderlippe@gmail.com>

TimvdLippe commented:
https://github.com/TimvdLippe/servo/actions/runs/15809517835

mrobinson commented on components/script/dom/csp.rs:
```diff
@@ -0,0 +1,305 @@
+    }
+}
+
+/// Used to determine which inline check to run
+pub use content_security_policy::Violation;
```
Please move this to the top of the file.

mrobinson commented on components/script/dom/csp.rs:
```diff
@@ -0,0 +1,305 @@
+pub(crate) fn parse_csp_list_from_metadata(headers: &Option<Serde<HeaderMap>>) -> Option<CspList> {
+    // TODO: Implement step 1 (local scheme special case)
+    let headers = headers.as_ref()?;
+    let mut csp = headers.get_all("content-security-policy").iter();
+    // This silently ignores the CSP if it contains invalid Unicode.
+    // We should probably report an error somewhere.
+    let c = csp.next().and_then(|c| c.to_str().ok())?;
+    let mut csp_list = CspList::parse(c, PolicySource::Header, PolicyDisposition::Enforce);
+    for c in csp {
+        let c = c.to_str().ok()?;
+        csp_list.append(CspList::parse(
+            c,
+            PolicySource::Header,
+            PolicyDisposition::Enforce,
+        ));
+    }
+    let csp_report = headers
+        .get_all("content-security-policy-report-only")
+        .iter();
+    // This silently ignores the CSP if it contains invalid Unicode.
+    // We should probably report an error somewhere.
+    for c in csp_report {
+        let c = c.to_str().ok()?;
+        csp_list.append(CspList::parse(
+            c,
+            PolicySource::Header,
+            PolicyDisposition::Report,
+        ));
+    }
+    Some(csp_list)
```
I think it's a bit confusing to have both `csp` and `c` variables here, so please use names that are more descriptive. Perhaps "csp_header" and "csp" or "policy" work better here.

TimvdLippe replied:
I copied the method as-is. When we refactor this code to handle headers properly, we can pick this up? I rather avoid method rewriting when copying, if that's possible.

mrobinson commented on components/script/dom/element.rs:
```diff
@@ -2259,15 +2259,19 @@ impl Element {
-                        source,
-                    ) == csp::CheckResult::Blocked
+                    if global
+                        .get_csp_list()
```
This doesn't need to be fixed in this change, but `fn get_csp_list()` does not follow rust naming conventions, which we try to follow in Servo. (see https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter). It should just be `fn csp_list()`.

mrobinson reviewed (approved):
Looks good, with some minor nits. Feel free to land when they are addressed. Thank you!

TimvdLippe committed:
Move violation use to top of file

Signed-off-by: Tim van der Lippe <TimvdLippe@users.noreply.github.com>

TimvdLippe commented:
@mrobinson I don't have merge rights on this repository. Do you mind to do the honors for me and click the merge button?
